src/java.base/unix/native/criuengine/criuengine.c

```diff
@@ -99,50 +168,69 @@ static int checkpoint(pid_t jvm,
-        *arg++ = log_file != NULL ? log_file : "dump4.log";
+    const char* args[32] = {
+        criu,
+        "dump",
+        "-t", jvmpidchar,
+        "-D", imagedir,
+        "--shell-job",
+    };
+    const char** arg = args + 7;
 
-        if (leave_running) {
-            *arg++ = "-R";
-        }
+    *arg++ = verbosity != NULL ? verbosity : "-v4";
+    *arg++ = "-o";
+    // -D without -W makes criu cd to image dir for logs
+    const char *log_local = log_file != NULL ? log_file : "dump4.log";
+    *arg++ = log_local;
 
-        char *criuopts = getenv("CRAC_CRIU_OPTS");
-        if (criuopts) {
-            char* criuopt = strtok(criuopts, " ");
-            while (criuopt && ARRAY_SIZE(args) >= (size_t)(arg - args) + 1/* account for trailing NULL */) {
-                *arg++ = criuopt;
-                criuopt = strtok(NULL, " ");
-            }
-            if (criuopt) {
-                fprintf(stderr, "Warning: too many arguments in CRAC_CRIU_OPTS (dropped from '%s')\n", criuopt);
-            }
+    if (leave_running) {
+        *arg++ = "-R";
+    }
+
+    char *criuopts = getenv("CRAC_CRIU_OPTS");
+    if (criuopts) {
+        char* criuopt = strtok(criuopts, " ");
+        while (criuopt && ARRAY_SIZE(args) >= (size_t)(arg - args) + 1/* account for trailing NULL */) {
+            *arg++ = criuopt;
+            criuopt = strtok(NULL, " ");
+        }
+        if (criuopt) {
+            fprintf(stderr, "Warning: too many arguments in CRAC_CRIU_OPTS (dropped from '%s')\n", criuopt);
         }
-        *arg++ = NULL;
+    }
+    *arg++ = NULL;
 
+    pid_t child = fork();
+    if (!child) {
         execv(criu, (char**)args);
-        perror("criu dump");
-        exit(1);
+        fprintf(stderr, "Cannot execute CRIU \"%s\": %m\n", criu);
+        exit(77);
     }
 
+    char *criu_cmdline = NULL;
+    char *criu_log = NULL;
     int status;
-    if (child != wait(&status) || !WIFEXITED(status) || WEXITSTATUS(status)) {
+    if (child != wait(&status)) {
+        criu_cmdline = join_args(args);
+        fprintf(stderr, "Error waiting for spawned CRIU \"%s\": %m\n", criu_cmdline);
+        kickjvm(jvm, -1);
+    } else if (!WIFEXITED(status)) {
+        criu_cmdline = join_args(args);
+        criu_log = path_abs2(imagedir, log_local);
+        fprintf(stderr, "Spawned CRIU \"%s\" has not properly exited: waitpid status %d - check %s\n", criu_cmdline, status, criu_log);
+        kickjvm(jvm, -1);
+    } else if (WEXITSTATUS(status)) {
+        if (WEXITSTATUS(status) != 77) {
+            criu_cmdline = join_args(args);
+            criu_log = path_abs2(imagedir, log_local);
+            fprintf(stderr, "Spawned CRIU \"%s\" has not properly exited: exit code %d - check %s\n", criu_cmdline, WEXITSTATUS(status), criu_log);
+        }
         kickjvm(jvm, -1);
     } else if (leave_running) {
         kickjvm(jvm, 0);
     }
+    free(criu_cmdline);
+    free(criu_log);
```

Comment:
Extensive use of free in the short living application is questionable. By dropping this we'll make it simpler and faster.